### PR TITLE
Appengine source validation

### LINKED ISSUE #1469: Quick fix shortcut (Ctrl+1) doesn't work for appengine-web.xml Project ID validation
I expect that when I press Ctrl+1 while the cursor is on the offending element, it should list the fix about the warning.

selection_009

### PR DESCRIPTION
**Pull request by abbiefarr**
fix #1469 @elharo @meltsufin 

PR does a couple things: 
  1) Provides source (as-you-type) validation for appengine-web-xml
  2) Provides a source (ctrl-1) quick fix

### CODE REVIEW

**Comment by abbiefarr** [comment]
My tests all pass locally but fail in Travis wherever `PlatformUI.getWorkbench()` is called (like in `ApplicationQuickAssistProcessorTest`). Not sure how to work around this without making the test worthless.

File: plugins/com.google.cloud.tools.eclipse.appengine.validation/src/com/google/cloud/tools/eclipse/appengine/validation/AbstractXmlSourceValidator.java [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+    } catch (IOException | CoreException | ParserConfigurationException ex) {
+      logger.log(Level.SEVERE, ex.getMessage());
+    }
+  }
+  
+  /**
+   * Adds an {@link IMessage} to XML file for every 
+   * {@link BannedElement} found in the file.
+   */
+  abstract protected void validate(IReporter reporter, byte[] bytes)
```
Comment (elharo) [inline plugins/com.google.cloud.tools.eclipse.appengine.validation/src/com/google/cloud/tools/eclipse/appengine/validation/AbstractXmlSourceValidator.java]: I think we usually write this as

protected abstract

File: plugins/com.google.cloud.tools.eclipse.appengine.validation/src/com/google/cloud/tools/eclipse/appengine/validation/AbstractXmlSourceValidator.java [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+      String encoding = getDocumentEncoding(document);
+      byte[] bytes = document.get().getBytes(encoding);
+      this.validate(reporter, bytes);
+    } catch (IOException | CoreException | ParserConfigurationException ex) {
+      logger.log(Level.SEVERE, ex.getMessage());
+    }
+  }
+  
+  /**
+   * Adds an {@link IMessage} to XML file for every
```
Comment (elharo) [inline plugins/com.google.cloud.tools.eclipse.appengine.validation/src/com/google/cloud/tools/eclipse/appengine/validation/AbstractXmlSourceValidator.java]: the XML file

File: plugins/com.google.cloud.tools.eclipse.appengine.validation/src/com/google/cloud/tools/eclipse/appengine/validation/XsltSourceQuickFix.java [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+  private String xsltPath;
+  private String message;
+  
+  public XsltSourceQuickFix(String xsltPath, String message) {
+    this.xsltPath = xsltPath;
+    this.message = message;
+  }
+
+  /**
+   * Applies the provided Xslt style sheet to the {@link IDocument} in the source editor.
```
Comment (elharo) [inline plugins/com.google.cloud.tools.eclipse.appengine.validation/src/com/google/cloud/tools/eclipse/appengine/validation/XsltSourceQuickFix.java]: XSLT

File: plugins/com.google.cloud.tools.eclipse.appengine.validation/src/com/google/cloud/tools/eclipse/appengine/validation/ApplicationQuickAssistProcessor.java [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+          return new ICompletionProposal[] {proposal};
+        }
+      } 
+    }
+    return null;
+  }
+  
+  @Override
+  public String getErrorMessage() {
+    return Messages.getString("no.fixes.found");
```
Comment (elharo) [inline plugins/com.google.cloud.tools.eclipse.appengine.validation/src/com/google/cloud/tools/eclipse/appengine/validation/ApplicationQuickAssistProcessor.java]: Do we actually show an error message if no quick fix is found. I thought Eclipse simply didn't offer a quick fix in that case.

Comment (abbiefarr) [inline plugins/com.google.cloud.tools.eclipse.appengine.validation/src/com/google/cloud/tools/eclipse/appengine/validation/ApplicationQuickAssistProcessor.java]: Yeah looks like Eclipse implementations just return null. Will change

**Comment by chanseokoh** [comment]
> My tests all pass locally but fail in Travis wherever PlatformUI.getWorkbench() is called (like in ApplicationQuickAssistProcessorTest). Not sure how to work around this without making the test worthless.

Add below to `.validation.test/pom.xml` (you already have `tycho-surefire-plugin`, so add only `<configuration>` for that), and at least it will work:

```xml
      <plugin>
        <groupId>org.eclipse.tycho</groupId>
        <artifactId>tycho-surefire-plugin</artifactId>
        <configuration>
          <useUIHarness>true</useUIHarness>
          <useUIThread>true</useUIThread>
          <product>org.eclipse.platform.ide</product>
        </configuration>
      </plugin>
      <plugin>
        <groupId>org.eclipse.tycho</groupId>
        <artifactId>target-platform-configuration</artifactId>
        <configuration>
          <dependency-resolution>
            <extraRequirements>
              <requirement>
                <type>eclipse-feature</type>
                <id>org.eclipse.rcp</id>
                <versionRange>0.0.0</versionRange>
              </requirement>
            </extraRequirements>
          </dependency-resolution>
        </configuration>
      </plugin>
```

**Comment by chanseokoh** [comment]
`PlatformUI.workbench()` will work, but a test will fail for a different reason. We can look into it later.

**Comment by chanseokoh** [comment]
I figured out adding the following plugin in `<extraRequirements>` will make it work.

```xml
              <requirement>
                <type>eclipse-plugin</type>
                <id>org.eclipse.wst.xml.ui</id>
                <versionRange>0.0.0</versionRange>
              </requirement>
```

However, then the test seems to timeout after 600s on Travis (albeit not locally, which is strange). Not sure why.

**Comment by chanseokoh** [comment]
Finally, loading `org.eclipse.platform` instead of `org.eclipse.rcp` made it work (although I still don't get why `mvn verify` succeeds locally with `org.eclipse.rcp`.)

Anyway, below are the changes needed. You can verify the tests have passed.

https://github.com/GoogleCloudPlatform/google-cloud-eclipse/compare/appengine_source_validation...chanseok-validation-test-2?expand=1

File: plugins/com.google.cloud.tools.eclipse.appengine.validation.test/src/com/google/cloud/tools/eclipse/appengine/validation/AppEngineWebXmlSourceValidatorTest.java [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+      "<test></test>".getBytes(StandardCharsets.UTF_8);
+  private static final byte[] XML =
+      "<application></application>".getBytes(StandardCharsets.UTF_8);
+  private IncrementalReporter reporter = new IncrementalReporter(null /*progress monitor*/);
+  
+  @Test
+  public void testValidate_noBannedElements() throws CoreException, IOException, ParserConfigurationException {
+    AppEngineWebXmlSourceValidator validator = new AppEngineWebXmlSourceValidator();
+    validator.validate(reporter, XML_NO_BANNED_ELEMENTS);
+    assertEquals(0, reporter.getMessages().size());
```
Comment (elharo) [inline plugins/com.google.cloud.tools.eclipse.appengine.validation.test/src/com/google/cloud/tools/eclipse/appengine/validation/AppEngineWebXmlSourceValidatorTest.java]: or just 

assertTrue(reporter.getMessages().isEmpty())

File: plugins/com.google.cloud.tools.eclipse.appengine.validation.test/src/com/google/cloud/tools/eclipse/appengine/validation/AppEngineWebXmlSourceValidatorTest.java [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+  }
+  
+  @Test
+  public void testValidate() throws CoreException, IOException, ParserConfigurationException {
+    AppEngineWebXmlSourceValidator validator = new AppEngineWebXmlSourceValidator();
+    validator.validate(reporter, XML);
+    assertEquals(1, reporter.getMessages().size());
+    reporter.removeAllMessages(validator);
```
Comment (elharo) [inline plugins/com.google.cloud.tools.eclipse.appengine.validation.test/src/com/google/cloud/tools/eclipse/appengine/validation/AppEngineWebXmlSourceValidatorTest.java]: add

assertTrue(reporter.getMessages().isEmpty())

at the end to make sure messages were removed

File: plugins/com.google.cloud.tools.eclipse.appengine.validation/src/com/google/cloud/tools/eclipse/appengine/validation/AbstractXmlSourceValidator.java [hunk 7]
```diff
@@ -0,0 +1,104 @@
...
+ * Abstract source view validator.
+ */
+public abstract class AbstractXmlSourceValidator implements ISourceValidator, IValidator {
+  IDocument document;
+  
+  private static final Logger logger = Logger.getLogger(
+      AbstractXmlSourceValidator.class.getName());
+  
+  /**
+   * Extracts byte[] from XML.
```
Comment (elharo) [inline plugins/com.google.cloud.tools.eclipse.appengine.validation/src/com/google/cloud/tools/eclipse/appengine/validation/AbstractXmlSourceValidator.java]: This comment doesn't seem to describe the main purpose of the method.

File: plugins/com.google.cloud.tools.eclipse.appengine.validation/src/com/google/cloud/tools/eclipse/appengine/validation/AppEngineWebXmlSourceValidator.java [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+   * Adds an {@link IMessage} to appengine-web.xml for every 
+   * {@link BannedElement} found in the file.
+   */
+  protected void validate(IReporter reporter, byte[] bytes) 
+      throws CoreException, IOException, ParserConfigurationException {
+    try {
+      SaxParserResults parserResults = BlacklistSaxParser.readXml(bytes);
+      Map<BannedElement, Integer> bannedElementOffsetMap =
+          ValidationUtils.getOffsetMap(bytes, parserResults);
+      String markerId = "com.google.cloud.tools.eclipse.appengine.validation.appEngineBlacklistMarker";
```
Comment (elharo) [inline plugins/com.google.cloud.tools.eclipse.appengine.validation/src/com/google/cloud/tools/eclipse/appengine/validation/AppEngineWebXmlSourceValidator.java]: Make markerId a final static constant field.

**Pull request merged**